File: backend/bootstrappers/javascript/wrapper.js
#!/usr/bin/env node
/**
 * Generic JavaScript bootstrapper.
 *
 * Invoked as: node wrapper.js <bot_file.js>
 *
 * The user's file must export a class named GameBot:
 *   class GameBot {
 *     constructor(playerId) { ... }
 *     makeMove(state) { ... }   // sync or async
 *   }
 *   module.exports = { GameBot };
 *
 * The wrapper reads the game name from INIT, loads the matching adapter,
 * and drives the stdin/stdout loop. Adapter types (Move, TicTacToeState,
 * etc.) are available to the user via require('./tictactoe') inside the
 * bot file or simply from the injected globals (see adapter injection below).
 */
"use strict";

const path = require("path");
const readline = require("readline");

const botFilePath = process.argv[2];
if (!botFilePath) {
  process.stderr.write("usage: node wrapper.js <bot_file.js>\n");
  process.exit(1);
}

const { GameBot } = require(path.resolve(botFilePath));

let bot = null;
let adapter = null;
let playerId = null;

function loadAdapter(gameName) {
  // e.g. "tictactoe" -> /app/games/tictactoe/adapters/javascript.js
  const adapterPath = path.join(__dirname, "..", "..", "games", gameName, "adapters", "javascript.js");
  return require(adapterPath);
}

const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity, terminal: false });

// Sequential message queue — guarantees in-order processing even for async makeMove.
let queue = Promise.resolve();

rl.on("line", (line) => {
  line = line.trim();
  if (!line) return;

  let msg;
  try {
    msg = JSON.parse(line);
  } catch (e) {
    process.stderr.write(`[wrapper] bad JSON: ${e.message}\n`);
    return;
  }

  queue = queue.then(async () => {
    const { type } = msg;

    if (type === "INIT") {
      adapter = loadAdapter(msg.game);
      playerId = msg.player;
      bot = new GameBot(playerId);

    } else if (type === "MOVE") {
      if (!bot) {
        process.stderr.write("[wrapper] received MOVE before INIT\n");
        process.exit(1);
      }
      try {
        const state = adapter.fromJson(playerId, msg);
        const move = await Promise.resolve(bot.makeMove(state));
        process.stdout.write(JSON.stringify({ move: adapter.toJson(move) }) + "\n");
      } catch (e) {
        process.stderr.write(e.stack + "\n");
        process.exit(1);
      }

    } else if (type === "END") {
      process.exit(0);
    }
    // Unknown types ignored.
  });
});

rl.on("close", () => {
  queue.then(() => process.exit(0));
});

File: backend/bootstrappers/python/wrapper.py
#!/usr/bin/env python3
"""
Generic Python bootstrapper.

Invoked as: python3 wrapper.py <bot_file.py>

The user's file must define a class named GameBot:
    class GameBot:
        def __init__(self, player_id: int): ...
        def makeMove(self, state) -> Move: ...

The wrapper reads the game name from the INIT message, loads the
matching game adapter, injects its types into the user's module
namespace, and drives the stdin/stdout loop.
"""
import sys
import json
import importlib
import importlib.util
from importlib.machinery import SourceFileLoader
import traceback


def load_user_module(path: str):
    import os
    if not os.path.isfile(path):
        sys.stderr.write(f"[wrapper] bot file not found at {path!r}\n")
        sys.stderr.write(f"[wrapper] /bot contents: {os.listdir('/bot') if os.path.isdir('/bot') else 'no /bot dir'}\n")
        sys.exit(1)
    # Explicitly pass SourceFileLoader so the file extension is irrelevant.
    # The mounted file is named "user_bot" with no .py extension.
    spec = importlib.util.spec_from_file_location(
        "user_bot", path, loader=SourceFileLoader("user_bot", path)
    )
    mod = importlib.util.module_from_spec(spec)
    sys.modules["user_bot"] = mod
    spec.loader.exec_module(mod)
    return mod


def load_adapter(game_name: str):
    # e.g. "tictactoe" -> games.tictactoe.adapters.python
    return importlib.import_module(f"games.{game_name}.adapters.python")


def main():
    if len(sys.argv) < 2:
        sys.stderr.write("usage: wrapper.py <bot_file.py>\n")
        sys.exit(1)

    user_mod = load_user_module(sys.argv[1])
    bot = None
    adapter = None
    player_id = None

    for raw in sys.stdin:
        raw = raw.strip()
        if not raw:
            continue

        try:
            msg = json.loads(raw)
        except json.JSONDecodeError as e:
            sys.stderr.write(f"[wrapper] bad JSON: {e}\n")
            continue

        msg_type = msg.get("type")

        if msg_type == "INIT":
            adapter = load_adapter(msg["game"])
            player_id = msg["player"]
            # Inject adapter's public symbols into the user module so the
            # user can write `from tictactoe import Move, Cell` etc.
            for name in dir(adapter):
                if not name.startswith("_"):
                    setattr(user_mod, name, getattr(adapter, name))
            bot = user_mod.GameBot(player_id)

        elif msg_type == "MOVE":
            if bot is None:
                sys.stderr.write("[wrapper] received MOVE before INIT\n")
                sys.exit(1)
            try:
                state = adapter.from_json(player_id, msg)
                move = bot.makeMove(state)
                response = json.dumps({"move": adapter.to_json(move)})
                sys.stdout.write(response + "\n")
                sys.stdout.flush()
            except Exception:
                sys.stderr.write(traceback.format_exc())
                sys.exit(1)

        elif msg_type == "END":
            sys.exit(0)

        # Unknown message types are silently ignored (forward-compatibility).


if __name__ == "__main__":
    main()
